server/src/main/java/org/elasticsearch/search/aggregations/InternalAggregations.java

```diff
@@ -236,11 +236,34 @@ public static InternalAggregations topLevelReduce(List<InternalAggregations> agg
         }
         if (context.isFinalReduce()) {
             List<InternalAggregation> reducedInternalAggs = reduced.getInternalAggregations();
-            reducedInternalAggs = reducedInternalAggs.stream()
-                .map(agg -> agg.reducePipelines(agg, context, context.pipelineTreeRoot().subTree(agg.getName())))
-                .collect(Collectors.toCollection(ArrayList::new));
+            List<InternalAggregation> internalAggregations = null;
+            for (int i = 0; i < reducedInternalAggs.size(); i++) {
+                InternalAggregation agg = reducedInternalAggs.get(i);
+                InternalAggregation internalAggregation = agg.reducePipelines(
+                    agg,
+                    context,
+                    context.pipelineTreeRoot().subTree(agg.getName())
+                );
+                if (internalAggregation.equals(agg) == false) {
+                    if (internalAggregations == null) {
+                        internalAggregations = new ArrayList<>(reducedInternalAggs);
```

Comment:
No need to be tricky here IMO, copying the full list is likely similar in cost to doing anything tricky like copying the first I unchanged elements and then moving to a different loop or so, given how short these lists are.